### PR TITLE
Relax a constraint for Alternative's unite and separate methods from Monad to FlatMap

### PR DESCRIPTION
**Pull request by satorg**
The initial discussion is [here](https://discord.com/channels/632277896739946517/632278570512678923/888660638073749514).

### Rationale

- The `unite` and `separate` methods don't need a `Monad` actually to do their job, because they only use `flatMap` in addition to methods from `Alternative`. Therefore the `FlatMap[F]` constraint is just enough there.
- Moreover, both `Alternative` and `Monad` implement `Applicative`. That means the `Applicative` methods can come by two different paths which creates sort of a loophole for ambiguity:
  ```scala
  implicit val fooMonad: Monad[Foo] = ??? // assume Monad[Foo] is different from Alternative[Foo]
  Alternative[Foo].unite[Bar, Int](Foo(Bar(123)) // which `pure` should it use – one from `Alternative` or `Monad` ? 
  ```

### Discussion

In fact, if it is okay to constraint the `unite` and `separate` methods with a `Monad`, then the `Alternative` typeclass is absolutely redundant in such a case – we can simply put these two methods directly into `MonoidK`.

Another approach (perhaps even better one) could be a new typeclass (let's name it `Monadditive` for the sake of this example) which would inherit to `Monad` and `MonoidK` (via `Alternative`) instead of just `Applicative` and `MonoidK` (which the current `Alternative` inherits to). In this case the definition of those two methods – `unite` and `separate` – could be simplified to the following:
```scala
trait Monadditive[F[_]] extends Alternative[F] with Monad[F] { // implies `MonoidK` as well
  def unite[G[_], A](fga: F[G[A]])(implicit G: Foldable[G]): F[A] = ??? // implementation remains the same
  def separate[G[_, _], A, B](fgab: F[G[A, B]])(implicit G: Bifoldable[G]): (F[A], F[B]) = ??? // implementation remains the same
}
```
In other words, an additional constraint for `F[_]` is not necessary anymore.

Therefore such a new typeclass would allow to reduce number of additional constraints passed to those methods.
And looks like it would make a lot of sense anyway – most of the containers which Cats supports implement corresponding `Monad` instances, not only `Applicative`.

### CODE REVIEW

**Comment by satorg** [comment]
Converting the PR to Draft since Mime is not happy with this it:
```
[error] (coreJVM / mimaReportBinaryIssues) Failed binary compatibility check against org.typelevel:cats-core_2.13:2.6.1! Found 1 potential problems
```
Let me investigate the issue first. If it is not worth the effort to fix it, I'll close the PR.

**Comment by rossabaker** [comment]
If you keep the old one and make it `private[cats]`, you'll fix that MiMa complaint.  Then you'll be fighting ambiguous overloads at call sites within cats, and cats-rooted projects like cats-effect.  You can fix the ambiguous overload by putting the old one in a bincompat parent trait, but then a different MiMa error pops up.  We've done this successfully on instances, but I don't remember a precedent on typeclasses.

There's also syntax to wrestle with.

**Comment by satorg** [comment]
@rossabaker thank you for the advice. As I can see now it won't be an easy lift. Seems I managed to calm MiMa down at this point, although the solution I came up with looks quite ugly to me. As you can see in the last commit, I added a new `AlternativeBinCompat0` trait for that. Initially I tried to move the old `unite` method to this trait, but it didn't make MiMa happy unfortunately. It simply changed the way it complaints to
```
[error] Cats core: Failed binary compatibility check against org.typelevel:cats-core_2.13:2.6.1! Found 1 potential problems
[error]  * in current version, classes mixing cats.Alternative need be recompiled to wire to the new static mixin forwarder method all super calls to method unite(java.lang.Object,cats.Monad,cats.Foldable)java.lang.Object
[error]    filter with: ProblemFilters.exclude[NewMixinForwarderProblem]("cats.Alternative.unite")
```
So I decided to try the opposite: place the new `unite` into `AlternativeBinCompat0` while leaving the old one in place (but restricting its visibility via `private[cats]`. Currently all checks pass, but I haven't got started with the syntax yet.

File: tests/src/test/scala/cats/tests/AlternativeSuite.scala [hunk 1]
```diff
@@ -9,7 +9,9 @@ class AlternativeSuite extends CatsSuite {
     forAll { (list: List[Option[String]]) =>
       val expected = list.collect { case Some(s) => s }
 
-      assert(Alternative[List].unite(list) === expected)
+      assert(
+        (Alternative[List]: AlternativeBinCompat0[List]).unite(list) === expected
```
Comment (satorg) [inline tests/src/test/scala/cats/tests/AlternativeSuite.scala:13]: To make sure the new version is used here.

File: core/src/main/scala/cats/Alternative.scala [hunk 2]
```diff
@@ -89,6 +88,16 @@ import scala.annotation.implicitNotFound
...
 }
 
+private[cats] trait AlternativeBinCompat0[F[_]] { self: Alternative[F] =>
+
+  def unite[G[_], A](fga: F[G[A]])(implicit FM: FlatMap[F], G: Foldable[G]): F[A] =
+    unite0(fga)
+
+  // Allows to share implementation between the old and new `unite` method versions.
+  final protected def unite0[G[_], A](fga: F[G[A]])(implicit FM: FlatMap[F], G: Foldable[G]): F[A] =
+    FM.flatMap(fga) { G.foldMapK(_)(pure)(self) }
```
Comment (satorg) [inline core/src/main/scala/cats/Alternative.scala:98]: I tried to change implementation from
```
FM.flatMap { a =>
  G.foldLeft(ga, empty[A])((acc, a) => combineK(acc, pure(a)))
}
```
to just
```
FM.flatMap(fga) { G.foldMapK(_)(pure)(self) }
```
It should be equivalent to the previous one. But can be even more efficient since `foldMapK` has optimized implementations at least for some of supported collections.

Comment (satorg) [reply core/src/main/scala/cats/Alternative.scala:98]: ... but if there're any concerns or objections with this approach, I won't resist to switch back to the original implementation.

**Comment by satorg** [comment]
Seems I managed to get through all the thorns with this task. The only caveat: Simulacrum Scalafix rules don't work correctly with all these binary compatibility tricks in `Alternative` I had to apply. But that Simulacrum generated syntax is not currently in use for `Alternative` anyway.

File: core/src/main/scala/cats/Alternative.scala [hunk 3]
```diff
@@ -15,39 +24,44 @@ import scala.annotation.implicitNotFound
...
    * scala> Alternative[List].unite(x)
    * res0: List[Int] = List(1, 2, 3, 4)
    * }}}
    */
-  def unite[G[_], A](fga: F[G[A]])(implicit FM: Monad[F], G: Foldable[G]): F[A] =
-    FM.flatMap(fga) { ga =>
-      G.foldLeft(ga, empty[A])((acc, a) => combineK(acc, pure(a)))
-    }
+  def unite[G[_], A](fga: F[G[A]])(implicit FM: FlatMap[F], G: Foldable[G]): F[A] =
+    FM.flatMap(fga) { G.foldMapK(_)(pure)(self) }
```
Comment (satorg) [inline core/src/main/scala/cats/Alternative.scala:33]: The original implementation is replaced with this better optimized one which utilizes `Foldable.foldMapK` (instead of the former's `foldLeft`).

**Comment by satorg** [comment]
cc @rossabaker, in case if you're curious in how I've fought it :)

**Comment by satorg** [comment]
Hmm.. Seems the build failure is not related to the PR:
```
[error] org.scalajs.ir.IRVersionNotSupportedException: Failed to deserialize a file compiled with Scala.js 1.7 (supported up to: 1.5): /home/runner/.cache/coursier/v1/https/repo1.maven.org/maven2/org/scala-js/scalajs-library_2.13/1.7.1/scalajs-library_2.13-1.7.1.jar:/scala/volatile.sjsir
[error] You may need to upgrade the Scala.js sbt plugin to version 1.7 or later.
[error] 	at org.scalajs.sbtplugin.ScalaJSPluginInternal$.enhanceIRVersionNotSupportedException(ScalaJSPluginInternal.scala:83)
[error] 	at org.scalajs.sbtplugin.ScalaJSPluginInternal$.$anonfun$scalaJSStageSettings$19(ScalaJSPluginInternal.scala:227)
```
I wonder – is the main branch broken (how come?)

**Comment by rossabaker** [comment]
Yes, that scala.js issue came up on another branch.  Pretty sure it's a result of #4018.

**Comment by rossabaker** [comment]
#4048 should fix the build.

File: core/src/main/scala/cats/syntax/alternative.scala [hunk 4]
```diff
@@ -13,11 +13,20 @@ trait AlternativeSyntax {
     new GuardOps(b)
 }
 
-final class UniteOps[F[_], G[_], A](private val fga: F[G[A]]) extends AnyVal {
+final class UniteOps[F[_], G[_], A] private[syntax] (protected val fga: F[G[A]])
+    extends AnyVal
+    with UniteOpsBinCompat0[F, G, A] {
+
+  @deprecated("use a FlatMap-constrained instead", "2.6.2")
```
Comment (rossabaker) [inline core/src/main/scala/cats/syntax/alternative.scala:20]: For consistency with your other deprecation messages:

```suggestion
  @deprecated("use a FlatMap-constrained version instead", "2.6.2")
```

**Review by rossabaker — COMMENTED** [review]
Nice work!  I've used package private to hide overloads, but can't recall ever using a protected.  It seems to work.  Have you tried calling it anywhere with something that has a `FlatMap` but no `Monad`?  I'm mostly sure it will still work, but want to be 100% sure.  (Maybe this encoding isn't as novel as I think.)

**Comment by satorg** [comment]
> I've used package private to hide overloads, but can't recall ever using a protected. It seems to work.

Right, good catch. First I tried to make it `private[cats]` indeed. And it works in terms of binary compatibility too. But it makes testing the new methods a way more sophisticated since apparently all the tests are also located inside the `cats` package and the compiler tries to route all the calls to `unite` and `separate`  to the old methods instead of the new ones. Using `protected` instead of `private[cats]` basically solves the problem while preserving the binary compatibility.

Interesting observation: this trick with `protected` does not work for syntax – it breaks BC for Scala 2.12 (but still works for Scala 2.13 and Scala 3). I don't know exactly why.

**Comment by rossabaker** [comment]
> Interesting observation: this trick with protected does not work for syntax – it breaks BC for Scala 2.12 

Oh, so this is still going to fail MiMa even after #4048 is merged? 🤔

**Comment by satorg** [comment]
> Oh, so this is still going to fail MiMa even after #4048 is merged? 🤔

No, I don't think so. I mean, all the checks had been succeeding before #4018. But later I rebased this PR onto `main` to test it against the latest version – and the checks began failing. I'm going to make the changes you requested and we will see what happens anyway :)

File: core/src/main/scala/cats/syntax/alternative.scala (hunk 4 above)
Comment (satorg) [reply core/src/main/scala/cats/syntax/alternative.scala]: Fixed.

**Comment by satorg** [comment]
> Have you tried calling it anywhere with something that has a FlatMap but no Monad?

Good catch, thanks. Seems I overlooked that it has to be explicitly tested. Should be good now.

**Review by rossabaker — APPROVED** [review]

File: core/src/main/scala/cats/syntax/alternative.scala [hunk 5]
```diff
@@ -13,11 +13,20 @@ trait AlternativeSyntax {
     new GuardOps(b)
 }
 
-final class UniteOps[F[_], G[_], A](private val fga: F[G[A]]) extends AnyVal {
```
Comment (satorg) [inline core/src/main/scala/cats/syntax/alternative.scala:16]: **Note**: the `fga`'s access is changed from `private` to `protected` because `fga` needs to be accessible from `UniteOpsBinCompat0`. The same change is made for the `SeparateOps` below.

**One more note**: the `UniteOps`'s constructor is made `private[syntax]`. That makes sense in general – I believe it is better to keep its access scope as narrow as possible. And apparently it does not affect the binary compatibility. But if someone has a concern regarding this change – speak up please, we can change this access back to `public`.

Comment (rossabaker) [reply core/src/main/scala/cats/syntax/alternative.scala:16]: The `fga` seems fine.

Changing the constructor visibility is a source-breaking change that's highly unlikely to break anybody's source.  I'd probably leave it public to be conservative in Cats, but I'm not too fussed either way.

**Review by rossabaker — DISMISSED** [review]

**Comment by satorg** [comment]
@rossabaker thank you for keeping eye on this PR and helping me with polishing it, I appreciated it.

Yeah, as you might notice, I eventually decided to revert  access for `UniteOps` and `SeparateOps` constructors back to public. Although I still believe it would be better to make them private to `cats`, but the improvement would be so minutiae then it wouldn't be worth the risk it imposes.

**Comment by satorg** [comment]
Ironically, this PR had got conflicts with another my PR that have been merged recently.

**Review by rossabaker — APPROVED** [review]

**Review by johnynek — APPROVED** [review]
